actors/market/src/lib.rs

```diff
@@ -543,137 +550,226 @@ impl Actor {
         let curr_epoch = rt.curr_epoch();
 
         let (activations, batch_ret) = rt.transaction(|st: &mut State, rt| {
+            let proposals = st.load_proposals(rt.store())?;
+            let states = st.load_deal_states(rt.store())?;
+            let pending_deals = st.load_pending_deals(rt.store())?;
+            let mut pending_deal_allocation_ids =
+                st.load_pending_deal_allocation_ids(rt.store())?;
+
             let mut deal_states: Vec<(DealID, DealState)> = vec![];
             let mut batch_gen = BatchReturnGen::new(params.sectors.len());
             let mut activations: Vec<SectorDealActivation> = vec![];
             let mut activated_deals: HashSet<DealID> = HashSet::new();
-            let mut sector_deals: Vec<(SectorNumber, SectorDealIDs)> = vec![];
-
-            for p in params.sectors {
-                let proposal_array = st.get_proposal_array(rt.store())?;
-
-                if p.deal_ids.iter().any(|id| activated_deals.contains(id)) {
-                    log::warn!(
-                        "failed to activate sector containing duplicate deals {:?}",
-                        p.deal_ids
-                    );
-                    batch_gen.add_fail(ExitCode::USR_ILLEGAL_ARGUMENT);
-                    continue;
-                }
+            let mut sectors_deals: Vec<(SectorNumber, SectorDealIDs)> = vec![];
 
-                let proposals = match get_proposals(&proposal_array, &p.deal_ids, st.next_id) {
-                    Ok(proposals) => proposals,
-                    Err(e) => {
-                        log::warn!("failed to get proposals for deals {:?}: {:?}", p.deal_ids, e);
-                        batch_gen.add_fail(e.exit_code());
-                        continue;
+            'sector: for p in params.sectors {
+                let mut verified_infos = Vec::new();
+                let mut nonverified_deal_space: BigInt = BigInt::zero();
+                for deal_id in &p.deal_ids {
+                    // Check each deal is present only once, within and across sectors.
+                    if activated_deals.contains(deal_id) {
```

Comment:
The activated_deals set is scoped outside the loop over sectors, so this is deduping over everything. It's tested (please verify).

Reply:
> so this is deduping over everything

I agree for other sectors, I'm concerned about within the same sector.  I think we are no longer handling the case where a single sector has duplicate deal ids since we used to do this in `get_proposals` which is no longer called here.   I might be missing something, will keep looking for tests / write something quick to check this out further.

Reply:
Here's a fork of this branch with a test `fails_to_activate_single_sector_duplicate_deals` with different behavior on this branch compared to master: https://github.com/filecoin-project/builtin-actors/tree/test/duplicate-deal-ids-same-sector

We should include this test in this PR and an analogous one for SectorContentAdded with the expected behavior there.

Reply:
Ah I see, thank you. I initially had the insertion into `activated_deals` earlier, inside the one and only loop over deal IDs. Since moving it out for the sector-atomic changes, the insertion is too late.

I've now fixed that and added the test you proposed, thanks. The tests for SectorContentChanged already covered this, but I've rearranged them to more closely follow the BatchActivateDeals arrangement.